pkgs/stdenv/adapters.nix

```diff
@@ -238,4 +238,23 @@ rec {
         allowSubstitutes = false;
       });
     });
+
+
+  /* Modify a stdenv so that it builds binaries with the specified list of
+     compilerFlags appended and passed to the compiler.
+
+     This example would recompile every derivation on the system with
+     -funroll-loops and -O3 passed to each gcc invocation.
+
+     Example:
+       nixpkgs.overlays = [
+         (self: super: { stdenv = super.useTheseFlags { stdenv = super.stdenv; compilerFlags = [ "-funroll-loops" "-O3" ]; }; })
+       ];
+  */
+  useTheseFlags = { stdenv, compilerFlags }:
+    stdenv.override (old: {
+      mkDerivationFromStdenv = extendMkDerivationArgs old (args: {
+        NIX_CFLAGS_COMPILE = toString (args.NIX_CFLAGS_COMPILE or "") + " ${lib.concatStringsSep " " compilerFlags}";
+      });
+    });
```

Comment:
Not sure if the name is fitting or if this is already possible. Also some doc should be added.

Reply:
Has been changed to withCFlags, which I think is correct.